Source/WebKit/Shared/PurgeableCachedDecodedImage.cpp

```diff
@@ -0,0 +1,153 @@
+ * SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
+ * INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
+ * CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
+ * ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF
+ * THE POSSIBILITY OF SUCH DAMAGE.
+ */
+
+#include "config.h"
+#include "PurgeableCachedDecodedImage.h"
+
+#if ENABLE(PURGEABLE_DECODED_IMAGE_DATA_CACHE)
+
+#include "CacheableDecodedImage.h"
+#include "WebProcess.h"
+#include <WebCore/MemoryCache.h>
+#include <WebCore/Timer.h>
+#include <wtf/NeverDestroyed.h>
+
+namespace WebKit {
+
+static constexpr Seconds cleanupDelay { 2_s };
+
+class PurgeableCachedDecodedImageTimer {
+public:
+    static PurgeableCachedDecodedImageTimer& singleton();
+
+    void start();
+    void defer();
+    void stop();
+
+private:
+    static void timerFired();
+
+    std::optional<DeferrableOneShotTimer> m_timer;
+};
+
+PurgeableCachedDecodedImageTimer& PurgeableCachedDecodedImageTimer::singleton()
+{
+    static LazyNeverDestroyed<PurgeableCachedDecodedImageTimer> timer;
+    static std::once_flag onceFlag;
+    std::call_once(onceFlag, [] {
+        timer.construct();
+    });
+    return timer;
+}
+
+void PurgeableCachedDecodedImageTimer::start()
+{
+    if (!m_timer)
+        m_timer.emplace(&PurgeableCachedDecodedImageTimer::timerFired, cleanupDelay);
+
+    m_timer->restart();
+}
+
+void PurgeableCachedDecodedImageTimer::defer()
+{
+    if (m_timer)
+        m_timer->restart();
+}
+
+void PurgeableCachedDecodedImageTimer::stop()
+{
+    m_timer.reset();
+}
+
+void PurgeableCachedDecodedImageTimer::timerFired()
+{
+    WebProcess::singleton().releaseAllRemoteImageResources();
+    MemoryCache::singleton().destroyAndCacheDecodedDataForAllImages();
+}
+
+std::unique_ptr<PurgeableCachedDecodedImage> PurgeableCachedDecodedImage::create(ShareableBitmap& bitmap)
+{
+    auto memory = createVolatileCopyOfBitmap(bitmap);
+    if (!memory)
+        return nullptr;
+
+    return makeUnique<PurgeableCachedDecodedImage>(memory.releaseNonNull(), bitmap.size(), bitmap.configuration());
+}
+
+RefPtr<SharedMemory> PurgeableCachedDecodedImage::createVolatileCopyOfBitmap(ShareableBitmap& bitmap)
+{
+    auto memory = SharedMemory::allocate(bitmap.sizeInBytes());
+    if (!memory)
+        return nullptr;
+
+    memcpy(memory->data(), bitmap.data(), memory->size());
+    memory->setVolatile();
+    return memory;
+}
+
+PurgeableCachedDecodedImage::PurgeableCachedDecodedImage(Ref<SharedMemory>&& memory, const WebCore::IntSize& size, const ShareableBitmapConfiguration& configuration)
+    : m_memory(WTFMove(memory))
+    , m_size(size)
+    , m_configuration(configuration)
+{
+    PurgeableCachedDecodedImageTimer::singleton().defer();
+}
+
+PurgeableCachedDecodedImage::~PurgeableCachedDecodedImage() = default;
+
+RefPtr<NativeImage> PurgeableCachedDecodedImage::createNativeImage()
+{
+    if (!m_memory)
+        return nullptr;
```

Comment:
I think we should `ASSERT(m_memory);` here. If we find `m_memory` is null that means this function was called before and we tried to `m_memory->setNonVolatile()` but it was unsuccessful so we set m_memory to null. So what is the point to call this function again?